### DISCUSSION TITLE
Clickhouse backups

### DISCUSSION THREAD

**Discussion by vitobotta** [plausible/analytics#1226] [Self-Hosted Support] [answered]
Hi! I set up Plausible but I am having trouble backing up the Clickhouse database. I'm using https://github.com/AlexAkulov/clickhouse-backup for this since I couldn't find a better tool. 
The problem with this tool is that is only support MergeTree tables, so the schema_migration table, being of type TinyLog, is ignored, meaning than after restoring that table is created and left empty. So when the Plausible container starts, it tries to re-run the migrations which will fail because those tables etc already exist. 

So the question is.... how can I back up and restore the Clickhouse database used by Plausible?

Thanks!

**Comment by ukutaht** [comment, marked as answer]
Good question. Currently I still use full OS image based backups so it's not really an issue for me.

I don't see a quick and easy way to solve this. Maybe switching from TinyLog to MergeTree would work even though TinyLog is a much more appropriate table engine for migrations. This would require some work on the [clickhouse_ecto](https://github.com/plausible/clickhouse_ecto/blob/master/lib/clickhouse_ecto/migration.ex#L46) library and changing existing schema_migrations tables.

**Reply by vitobotta** [reply]
Hi @ukutaht and thanks for your reply! I wish I was familiar enough with Elixir to contribute. Do you just take snapshots of the disk/volume? If yes, isn't there the risk of inconsistent backups if writes happen right during the backup? Thanks

**Reply by ukutaht** [reply]
I was worried about this as well but I've now restored from our snapshots at least 5 times with no issues. Clickhouse inserts are atomic under certain conditions: https://github.com/ClickHouse/ClickHouse/issues/9195#issuecomment-587500824

This is clearly not the best way to handle it obviously. I'm planning to change our backup system to something better like `clickhouse-backup` later this year. So I'll be forced to fix it then.

**Reply by vitobotta** [reply]
Gotcha, for now I will just use Velero then to snapshot the volume then. I've also tested a restore once (with my very little data) and will do one more. Thanks! :)

**Reply by vitobotta** [reply]
Hi again :)

How do you take backups of the Clickhouse disk/volume? It is a regular backup or a snapshot at file system level?

I am doing backups of the Clickhouse volume with Velero because I use Kubernetes for hosting my stuff. Velero uses Restic under the hood to back up volumes, so it's a file-level backup. Sometimes these backups fail unfortunately, confirming my worries. So yeah, would be interesting to know what kind of backups you do. Thanks

**Reply by bradkane** [reply]
@ukutaht I assume that you are using Digital Ocean's snapshot feature.  Are you running your entire Plausible instance on one Droplet (including POSTGRES)?  Weekly backups are a standard (paid) feature, but how frequently are you snapshotting the server?  I guess what I am really asking about is a best practices guide to running a production-grade Plausible self-hosted instance.  And my guess is whatever you are doing IS the best current practice:)

**Reply by ukutaht** [reply]
@bradkane good questions

>  Are you running your entire Plausible instance on one Droplet (including POSTGRES)?

No, I have a dedicated droplet for Clickhouse. For postgres we use DO's managed postgres.

> Weekly backups are a standard (paid) feature, but how frequently are you snapshotting the server?

Currently I rely on standard weekly backups. I've tested a few and they seem to work OK.

Our setup is pretty much the same as when we launched the beta. I've been focusing on the product a lot and haven't had much time to improve our infrastructure. My plan is to focus on infrastructure stuff from October until the end of the year. I'll probably set up `clickhouse-copier` to take incremental backups and store them in object storage w/ disaster recovery protocols in place. Hopefully I can give more guidance then as to what proper production-grade Plausible infrastructure looks like.

**Reply by vitobotta** [reply]
> @bradkane good questions
> 
> > Are you running your entire Plausible instance on one Droplet (including POSTGRES)?
> 
> No, I have a dedicated droplet for Clickhouse. For postgres we use DO's managed postgres.
> 
> > Weekly backups are a standard (paid) feature, but how frequently are you snapshotting the server?
> 
> Currently I rely on standard weekly backups. I've tested a few and they seem to work OK.
> 
> Our setup is pretty much the same as when we launched the beta. I've been focusing on the product a lot and haven't had much time to improve our infrastructure. My plan is to focus on infrastructure stuff from October until the end of the year. I'll probably set up `clickhouse-copier` to take incremental backups and store them in object storage w/ disaster recovery protocols in place. Hopefully I can give more guidance then as to what proper production-grade Plausible infrastructure looks like.

I guess you use block storage volumes for Clickhouse? Or do you keep all the data on a single droplet? I would think you've got quite a bit of data by now in there.

**Reply by ukutaht** [reply]
> I guess you use block storage volumes for Clickhouse? Or do you keep all the data on a single droplet? I would think you've got quite a bit of data by now in there.

Not yet but it's something I'll be playing around with soon. Our droplet SSD is filling up fast.

**Reply by vitobotta** [reply]
How can you migrate the existing data to block storage without downtime? Can you create replicas on the new storage then promote those and delete the old ones? I'm not familiar with Clickhouse :)

**Reply by kees-closed** [reply]
I don't get why this comment is marked as an answer. Restoring a full system backup because you lost analytic metrics is messy. What if you also run other services on that host? Then you'll restore those to an earlier state as well and still lose data. Or what if you want to move the containers and data to a new machine? A full VM snapshot is great for a major issue. But it shouldn't be the answer for application backups.

**Reply by schnerring** [reply]
> The problem with this tool is that is only support MergeTree tables, so the schema_migration table, being of type TinyLog, is ignored, meaning than after restoring that table is created and left empty. 

I've been using `clickhouse-backup` to backup my Plausible data and just found out the hard way that my backups were incomplete.

When implementing my backup strategy, I was confident enough that my backups were fine and tested by restoring the Clickhouse DB _while Plausible was running_. It worked and Plausible kept running the past 2 months without any hiccups.

Things broke after Plausible was restarted due to an automatic Kubernetes node upgrade which restarted the all the pods. `/entrypoint.sh db migrate` failed because it tried re-applying all the migrations despite the DB already being migrated.

I'm so glad I pinned the Docker image version allowing me to re-create an empty Clickhouse DB and re-construct the `schema_migrations` table. Here's the query I used to fix the migrations:

**NOTE: THE DATA IS DIFFERENT DEPENDING ON THE PLAUSIBLE VERSION YOU'RE USING. THE FOLLOWING APPLIES TO v1.4.4**

```sql
INSERT INTO plausible.schema_migrations (version,inserted_at) VALUES
	 (20200915070607,'2022-04-22 04:07:40'),
	 (20200918075025,'2022-04-22 04:07:40'),
	 (20201020083739,'2022-04-22 04:07:40'),
	 (20201106125234,'2022-04-22 04:07:40'),
	 (20210323130440,'2022-04-22 04:07:40');
```

Things seem to work but could someone verify this is correct?

I think https://github.com/plausible/analytics/tree/v1.4.4/priv/repo/migrations contains both Clickhouse and Postgres migrations. But it seems this file list doesn't translate into a 1:1 mapping inside the `schema_migrations` tables in Postgres/Clickhouse.

**Reply by ukutaht** [reply]
Yes this seems correct to me. Indeed, `clickhouse-backup` only backs up the main tables, it doesn't back up the `schema_migrations` table. That has to be recovered manually at the moment.

**Reply by vitobotta** [reply]
> > The problem with this tool is that is only support MergeTree tables, so the schema_migration table, being of type TinyLog, is ignored, meaning than after restoring that table is created and left empty.
> 
> I've been using `clickhouse-backup` to backup my Plausible data and just found out the hard way that my backups were incomplete.
> 
> When implementing my backup strategy, I was confident enough that my backups were fine and tested by restoring the Clickhouse DB _while Plausible was running_. It worked and Plausible kept running the past 2 months without any hiccups.
> 
> Things broke after Plausible was restarted due to an automatic Kubernetes node upgrade which restarted the all the pods. `/entrypoint.sh db migrate` failed because it tried re-applying all the migrations despite the DB already being migrated.
> 
> I'm so glad I pinned the Docker image version allowing me to re-create an empty Clickhouse DB and re-construct the `schema_migrations` table. Here's the query I used to fix the migrations:
> 
> **NOTE: THE DATA IS DIFFERENT DEPENDING ON THE PLAUSIBLE VERSION YOU'RE USING. THE FOLLOWING APPLIES TO v1.4.4**
> 
> ```sql
> INSERT INTO plausible.schema_migrations (version,inserted_at) VALUES
> 	 (20200915070607,'2022-04-22 04:07:40'),
> 	 (20200918075025,'2022-04-22 04:07:40'),
> 	 (20201020083739,'2022-04-22 04:07:40'),
> 	 (20201106125234,'2022-04-22 04:07:40'),
> 	 (20210323130440,'2022-04-22 04:07:40');
> ```
> 
> Things seem to work but could someone verify this is correct?
> 
> I think https://github.com/plausible/analytics/tree/v1.4.4/priv/repo/migrations contains both Clickhouse and Postgres migrations. But it seems this file list doesn't translate into a 1:1 mapping inside the `schema_migrations` tables in Postgres/Clickhouse.

Hi @schnerring can you clarify how you fount the correct migrations from that long list? How to tell apart clickhouse migrations from postgres migrations? Thanks

**Reply by ruslandoga** [reply]
👋 @vitobotta 

ClickHouse migrations are defined in  https://github.com/plausible/analytics/tree/master/priv/clickhouse_repo/migrations

**Reply by vitobotta** [reply]
Thanks @ruslandoga

**Reply by schnerring** [reply]
The new location for future versions is: https://github.com/plausible/analytics/tree/master/priv/ingest_repo/migrations

Version v1.5.1 migrations can be found at the old location: https://github.com/plausible/analytics/tree/v1.5.1/priv/clickhouse_repo/migrations

**Reply by ruslandoga** [reply]
Yes, the migrations have been moved. Thank you for the update!

**Reply by schnerring** [reply]
Additionally, [setting `USE_EMBEDDED_BACKUP_RESTORE=true` with clickhouse-backup v2.0.0+](https://github.com/AlexAkulov/clickhouse-backup/releases/tag/v2.0.0) now supports backing up non-MergeTree tables.

However, it requires ClickHouse v23+ and is still experimental, see: https://github.com/AlexAkulov/clickhouse-backup/issues/238#issuecomment-1475127681

**Comment by bradkane** [comment]
@ukutaht I appreciate your input on this and understand that Q4 2021 is the infrastructure quarter.  I really like your idea of using DO's Managed POSTGRES solution.  And a clickhouse backup option to object storage makes perfect sense.  We are using that same strategy on another project (we send the objects to Amazon S3 rather than DO Spaces).  When we established this system, DO Spaces was very new and limited.  In the future, it might be helpful for others to share some metrics on production database sizes (POSTGRES, Clickhouse) for a given number of recorded events (assuming that is the right proxy for size).  Currently we have only a dev Plausible server (and yes, all on one droplet), but with the upcoming release of city level geo tracking, we will spin up a self-hosted production server as well.

**Reply by ukutaht** [reply]
I've actually hit very different bottlenecks with our system so far and for us it's quite a bit more complex than just giving some specs for certain number of events.

Early on some of our queries crashed the whole database because they would attempt to use more memory than was available on the machine. We gave it way more memory now so it doesn't happen. I believe Clickhouse also added some safeguards to make sure it runs OK on machines with <64GB ram.

Storage is obviously an ongoing concern. Luckily Clickhouse is very good at compression.

Most recently we started having problems with high CPU utilization which increased query latency to unacceptable territory.

I believe CPU util is not a huge concern for most self-hosters. See, it's very different running a single site with 100m monthly events vs 10,000 sites with 100m monthly events combined. This is because all the weekly reports, spike notifications and regular dashboard views add up to a lot of queries. If you're just self-hosting, you probably don't have the problem of trying to run 10,000 email reports on a Monday, each triggering at least 5 heavy DB queries :)

And because we run it very differently than self-hosters, it's not easy for me to provide specs that are relevant to you. Our requirements are quite different and I don't have much experience operating Plausible like self-hosters do.

**Reply by ukutaht** [reply]
I do run a small instance for self-hosting setup testing purposes. It's the $10/mo DO droplet with 1vCPU and 2GB ram. It cannot report one year's worth of stats from our live demo, for example, Clickhouse just crashes. I guess I can try bumping that up to see what sort of specs are needed to run a self-hosted instance that can handle our live demo.

**Reply by bradkane** [reply]
Thinking about this a little.  Are the heavy DB queries (when creating reports to be emailed) to the Clickhouse DB, POSTGRES DB or both?  Would separating Plausible into 3 droplets (Plausible core, Clickhouse, POSTGRES) help this situation?  As you know, droplets are very reasonable.  The POSTGRES droplet component could be DO's Managed POSTGRES for many benefits.  Can large accounts (2M monthly events) reside in the same self-hosted instance as 10 very small websites with perhaps 1000 monthly events each?

**Reply by ukutaht** [reply]
>  Are the heavy DB queries (when creating reports to be emailed) to the Clickhouse DB, POSTGRES DB or both? 

Just Clickhouse

> Would separating Plausible into 3 droplets (Plausible core, Clickhouse, POSTGRES) help this situation?

This is how I have it set up already. 

> Can large accounts (2M monthly events) reside in the same self-hosted instance as 10 very small websites with perhaps 1000 monthly events each?

I think if you're provisioned for a site with 2M monthly events then adding 10 tiny sites will not be a problem at all. The 'many small sites' problem we have is dealing with >10k sites. 10 is nothing :). In this scenario your bottleneck will be the one big site.

**Comment by vitobotta** [comment]
Hi all, has anything changed  / become easier since I started this discussion a while ago? Is there an easy way to properly back up the data now? Thanks @ukutaht
